src/core/internal/atomic.d

```diff
@@ -0,0 +1,734 @@
+                asm pure nothrow @nogc @trusted
+                {
+                    naked;
+                    push RBX;
+                    mov RAX, [RDX];
+                    mov RDX, 8[RDX];
+                    mov RBX, [RCX];
+                    mov RCX, 8[RCX];
+                    lock; cmpxchg16b [R8];
+                    setz AL;
+                    pop RBX;
+                    ret;
+                }
+            }
+            else
+            {
+                asm pure nothrow @nogc @trusted
+                {
+                    naked;
+                    push RBX;
+                    mov RAX, RDX;
+                    mov RDX, RCX;
+                    mov RBX, RDI;
+                    mov RCX, RSI;
+                    lock; cmpxchg16b [R8];
+                    setz AL;
+                    pop RBX;
+                    ret;
+                }
+            }
+        }
+    }
+    else
+        static assert (false, "Unsupported architecture.");
+}
+
+void atomicFence(MemoryOrder order = MemoryOrder.seq)() nothrow @nogc @safe
+{
+    // TODO: `mfence` should only be required for seq_cst operations, but this depends on
+    //       the compiler's backend knowledge to not reorder code inappropriately,
+    //       so we'll apply it conservatively.
+    static if (order != MemoryOrder.raw)
+    {
+        version (D_InlineAsm_X86)
+        {
+            import core.cpuid;
+
+            // TODO: review this implementation; it seems way overly complicated
+            asm pure nothrow @nogc @trusted
+            {
+                naked;
+
+                call sse2;
+                test AL, AL;
+                jne Lcpuid;
+
+                // Fast path: We have SSE2, so just use mfence.
+                mfence;
+                jmp Lend;
+
+            Lcpuid:
+
+                // Slow path: We use cpuid to serialize. This is
+                // significantly slower than mfence, but is the
+                // only serialization facility we have available
+                // on older non-SSE2 chips.
+                push EBX;
+
+                mov EAX, 0;
+                cpuid;
+
+                pop EBX;
+
+            Lend:
+
+                ret;
+            }
+        }
+        else version (D_InlineAsm_X86_64)
+        {
+            asm nothrow @nogc @trusted
+            {
+                naked;
+                mfence;
+                ret;
+            }
+        }
+    }
+    else
+        static assert (false, "Unsupported architecture.");
+}
+
+
+private:
+
+version (Windows)
+{
+    enum RegisterReturn(T) = is(T : U[], U) || is(T : R delegate(A), R, A...);
+}
+
+enum CanCAS(T) = is(T : ulong) ||
+                 is(T == class) ||
+                 is(T : U*, U) ||
+                 is(T : U[], U) ||
+                 is(T : R delegate(A), R, A...) ||
+                 (is(T == struct) && __traits(isPOD, T) &&
+                  T.sizeof <= size_t.sizeof*2 && // no more than 2 words
```

Comment:
This introduced a regression from the original:
```
T.sizeof <= 16
```
In that now atomics fail on ix86 and x32 platforms where 128-bit atomics is supported in the library (or natively with a `lock cmpxchg16b` on x32)